src/IR/DConToFunc.hs

```diff
@@ -114,9 +114,89 @@ createArityEnv defs = M.fromList $ concatMap arities defs
 dConToFunc :: I.Program Poly.Type -> Compiler.Pass (I.Program Poly.Type)
 dConToFunc p@I.Program { I.programDefs = defs, I.typeDefs = tDefs } =
   runArityFn (createArityEnv tDefs) $ do
-    defs' <- everywhereM (mkM dataToApp) defs
-    return p { I.programDefs = defs' ++ concat (createFuncs <$> tDefs) }
+    defs'' <- defs' -- user defined functions
+    return p { I.programDefs = tDefs' ++ defs'' } -- constructor funcs ++ user funcs
  where
+  tDefs' = concat (createFuncs <$> tDefs)
+  -- ^ top-level constructor functions
+  defs' =
+    filter ((`notElem` (fst <$> tDefs')) . fst)
+      <$> everywhereM (mkM dataToApp) defs
+  -- ^ We need to filter defs to account for name collisions
+  -- ^ between user defined funcs and newly created and inserted constructor funcs
+  {- Example of name collision case and correction
+  Given the SSLANG program
+  @
+  type Color
+    White
+    Black
+    RGB Int Int Int
+
+  main ( cout : & Int ) -> () = ()
+  @
+  dConToFunc will produce
+  @
+  type Color
+    White
+    Black
+    RGB Int Int Int
+
+  __RGB (__arg0 : Int) (__arg1 : Int) (__arg2 : Int) -> Color = 
+  RGB __arg0 __arg1 __arg2
+    
+  main ( cout : & Int ) -> () = ()
+  @
+  This is okay.
+  Now, given the SSLANG program
+  @
+  type Color
+    White
+    Black
+    RGB Int Int Int
+
+  __RGB (r : Int) (g : Int) (b : Int) -> Color = 
+  RGB r (g+2) b
+    
+  main ( cout : & Int ) -> () = ()
+  @
+  dConToFunc (without filtering defs!) produces
+  @
+  type Color
+    White
+    Black
+    RGB Int Int Int
+
+  __RGB (__arg0 : Int) (__arg1 : Int) (__arg2 : Int) -> Color = 
+  RGB __arg0 __arg1 __arg2
+
+  __RGB (r : Int) (g : Int) (b : Int) -> Color = 
+  RGB r (g+2) b
+    
+  main ( cout : & Int ) -> () = ()
+  @
+  This is not okay, because there are two functions named > __RGB.
+  To prevent this duplicate function case, we search defs for any user defined functions
+  that have the same name as our newly inserted constructor functions, and remove them.
```

Comment:
This seems like it would actually break programs that define functions that happen to clash with compiler-generated functions. Those functions should not be allowed in the first place.

Reply:
I agree that users should not be able to write functions that clash with compiler-generated functions. I remember we changed the scanner to allow users to begin function names with underscores around April or early May. Starting functions with underscores was the only way I knew of to distinguish compiler generated vs user defined functions. 
1) Would you like me to revert that change to the scanner? The issue here is if we do change the scanner, we can no longer read back in the output of the pretty printer, because dcontofunc generates functions that start with underscores. 
2) Alternatively, maybe dcontofunc could output its constructor functions to a different file that gets linked in. Then the output of the pretty printer could be tested (and fed back in to the front of the compiler), and if any functions with name collisions are detected by dcontofunc in the original source, it could throw an error.

Reply:
Let's fix this later! For now, I want to finalize this PR so that I can make some destructive changes on the IR (: